web/thesauruses/javascript/ECMAScript 2009/functions.json

```diff
@@ -0,0 +1,113 @@
+{
+  "meta": {
+    "language": "javascript",
+    "language_name": "JavaScript",
+    "structure": "functions",
+    "language_version": "ECMAScript 2009"
+  },
+  "concepts": {
+    "void_function_no_parameters": {
+      "name": "Function that does not return a value and takes no parameters",
+      "code": "function method() {}"
+    },
+    "void_function_with_parameters": {
+      "name": "Function that does not return a value and that takes 1 or more defined parameters",
+      "code": [
+        "function method(param1) {}",
+        "function method(param1, param2) {}"
+      ]
+    },
+    "void_function_with_keyword_parameters": {
+      "name": "Function that does not return a value and that takes 0 or more defined keyword parameters",
+      "code": [
+        "function method({param1}) {}",
+        "function method({param1, param2}) {}"
+      ]
+    },
+    "void_function_variable_parameters": {
+      "name": "Function that does not return a value and function that takes an unknown number of parameters",
+      "code": [
+        "function method() {",
+        "  var params = Array.prototype.slice.call(arguments); // convert arguments to an array",
```

Comment:
I think this is kind of "inventing" a solution? Typically I've seen this as something like `void functionName (int a, int b, int c...)` and `c` is the array of variable arguments. This looks like you're sort of hacking the function to grab some params somewhere. (also is `arguments` an actual thing? looks like an undefined variable)

Reply:
`arguments` is an object available inside a function.
But, I've used `slice` on `arguments` which is a bad idea according to the french version of MDN: "because it can prevent some JavaScript engine optimizations".
Here is their proposal to avoid using `slice` method:
```js
var args = (arguments.length === 1 ? [arguments[0]] : Array.apply(null, arguments));
```

References : 
- MDN page of `arguments` (english version): https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/arguments
- Warning about `slice` with `arguments` (french version): https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Functions/arguments#utilisation_de_la_d%C3%A9composition_avec_les_arguments

Reply:
For this, I'll let it stay since it's sort of a weird thing built into JS despite not having a direct "variable argument function" coorelation.